File: lib/proxy.js
import http  from 'http';
import https from 'https';
import { createStreamTransformer, convertAnthropicResponse, convertOpenAIChatResponse } from './converter.js';

// 流式传输活跃超时默认值：超过此时间没有新数据则终止连接（可由配置覆盖）
const DEFAULT_STREAM_ACTIVITY_TIMEOUT = 60_000; // 60 秒

/**
 * 转发请求到上游 provider，处理流式/非流式响应，并转换回客户端格式
 * @param {object} opts
 *   clientReq     - 原始客户端请求对象
 *   clientRes     - 原始客户端响应对象
 *   upstreamBody  - 已转换好的上游请求 JSON 对象
 *   provider      - { baseUrl, apiKey, format }
 *   clientFormat  - 'anthropic' | 'openai-chat' | 'openai-responses'
 *   originalModel - 客户端发来的原始模型名
 *   isStreaming    - 是否流式
 *   log           - 日志函数
 *   startTime     - 请求开始时间
 *   timeout       - 请求超时（毫秒）
 *   streamActivityTimeout - 流式活跃超时（毫秒），超过此时间无新数据则主动断开
 * @returns {Promise} 响应完成后 resolve（释放队列 slot）
 */
export function forwardRequest(opts) {
  const { clientReq, clientRes, upstreamBody, provider, clientFormat, originalModel, isStreaming, log, startTime, timeout = 60000, streamActivityTimeout = DEFAULT_STREAM_ACTIVITY_TIMEOUT } = opts;

  const upstreamUrl  = new URL(provider.baseUrl.startsWith('http') ? provider.baseUrl : `https://${provider.baseUrl}`);
  const basePath     = upstreamUrl.pathname.replace(/\/$/, '').replace(/\/v1$/, '');
  const isHttps      = upstreamUrl.protocol === 'https:';
  const upstreamPath = basePath + (provider.format === 'anthropic' ? '/v1/messages' : '/v1/chat/completions');

  const bodyStr = JSON.stringify(upstreamBody);

  const headers = {
    'content-type':   'application/json',
    'content-length': Buffer.byteLength(bodyStr),
    'accept':         'application/json, text/event-stream',
  };
  if (provider.format === 'anthropic') {
    headers['x-api-key']         = provider.apiKey;
    headers['anthropic-version'] = '2023-06-01';
  } else {
    headers['authorization'] = `Bearer ${provider.apiKey}`;
  }

  // 返回 Promise，响应完成后 resolve
  return new Promise((resolve, reject) => {
    // 流式活跃超时主动断开上游时置位，供 proxyReq error handler 识别，避免把预期内的保护动作误报为转发失败
    const streamCtx = { aborted: false };

    const proxyReq = (isHttps ? https : http).request({
      hostname: upstreamUrl.hostname,
      port:     upstreamUrl.port || (isHttps ? 443 : 80),
      path:     upstreamPath,
      method:   'POST',
      headers,
      timeout:  timeout,
    }, (proxyRes) => {
      const elapsed = startTime ? Date.now() - startTime : 0;

      // 精简的响应日志
      if (proxyRes.statusCode >= 400) {
        log(`← HTTP ${proxyRes.statusCode} [${elapsed}ms]`);
      } else {
        log(`← ${proxyRes.statusCode} [${elapsed}ms]`);
      }

      // 错误响应：立即释放 slot
      if (proxyRes.statusCode >= 400) {
        handleError(proxyRes, clientRes, proxyRes.statusCode, log);
        resolve();
        return;
      }

      // 非流式响应：响应到达即可释放
      if (!isStreaming) {
        handleResponse(proxyRes, clientRes, provider.format, clientFormat, originalModel, log);
        resolve();
        return;
      }

      // 流式响应：等待传输完成后再释放
      handleStream(proxyRes, clientRes, provider.format, clientFormat, originalModel, clientReq.headers, () => {
        resolve();
      }, log, streamCtx, streamActivityTimeout);
    });

    // 请求超时
    proxyReq.on('timeout', () => {
      proxyReq.destroy();
      const err = new Error('请求超时');
      log(`转发失败: ${err.message}`);
      if (!clientRes.headersSent) {
        clientRes.writeHead(504, { 'content-type': 'application/json' });
        clientRes.end(JSON.stringify({ error: { type: 'timeout_error', message: err.message } }));
      }
      reject(err);
    });

    // 请求错误
    proxyReq.on('error', (err) => {
      // 流式活跃超时主动断开上游：属于预期内的保护动作，handleStream 已打印明确日志并完成客户端收尾，这里不再重复误报
      if (streamCtx.aborted) {
        resolve();
        return;
      }
      log(`转发失败: ${err.message}`);
      if (!clientRes.headersSent) {
        clientRes.writeHead(502, { 'content-type': 'application/json' });
        clientRes.end(JSON.stringify({ error: { type: 'proxy_error', message: err.message } }));
      }
      reject(err);
    });

    proxyReq.write(bodyStr);
    proxyReq.end();
  });
}

// ── 非流式响应处理 ────────────────────────────

function handleResponse(proxyRes, clientRes, providerFormat, clientFormat, originalModel, log) {
  let raw = '';
  proxyRes.on('data', c => raw += c);
  proxyRes.on('end', () => {
    let data;
    try { data = JSON.parse(raw); }
    catch (e) {
      if (!clientRes.headersSent) {
        clientRes.writeHead(502, { 'content-type': 'application/json' });
        clientRes.end(JSON.stringify({ error: { type: 'parse_error', message: `上游响应解析失败: ${e.message}` } }));
      }
      return;
    }

    const result = providerFormat === 'anthropic'
      ? convertAnthropicResponse(data, clientFormat, originalModel)
      : convertOpenAIChatResponse(data, clientFormat, originalModel);

    const body = JSON.stringify(result);
    if (!clientRes.headersSent) {
      clientRes.writeHead(200, {
        'content-type':   'application/json',
        'content-length': Buffer.byteLength(body),
      });
    }
    clientRes.end(body);
  });

  // 上游响应流出错时，确保客户端收到响应，防止队列 slot 泄漏
  proxyRes.on('error', (err) => {
    log(`上游响应流错误: ${err.message}`);
    if (!clientRes.headersSent) {
      clientRes.writeHead(502, { 'content-type': 'application/json' });
      clientRes.end(JSON.stringify({ error: { type: 'upstream_error', message: err.message } }));
    } else if (!clientRes.writableEnded) {
      clientRes.end();
    }
  });
}

// ── 流式响应处理 ──────────────────────────────

function handleStream(proxyRes, clientRes, providerFormat, clientFormat, originalModel, reqHeaders, onDone, log, streamCtx, streamActivityTimeout = DEFAULT_STREAM_ACTIVITY_TIMEOUT) {
  let done = false;
  function finish() {
    if (done) return;
    done = true;
    clearTimeout(activityTimer);
    if (onDone) onDone();
  }

  // 向客户端补发一个合规的 SSE 收尾事件，避免上游中断时客户端只收到截断流
  function gracefulSseClose(message) {
    if (clientRes.writableEnded) return;
    try {
      if (clientFormat === 'anthropic') {
        // Anthropic 流式协议：中途异常以 error 事件结束
        clientRes.write(`event: error\n`);
        clientRes.write(`data: ${JSON.stringify({ type: 'error', error: { type: 'timeout_error', message } })}\n\n`);
      } else {
        // OpenAI Chat / Responses：以 [DONE] 标记流结束
        clientRes.write('data: [DONE]\n\n');
      }
    } catch { /* 客户端可能已断开，忽略写入异常 */ }
    clientRes.end();
  }

  // 活跃超时：上游停止发送数据超过指定时间未推送新数据，主动断开并优雅收尾
  function onActivityTimeout() {
    if (done) return;
    // 标记为主动断开，供 proxyReq error handler 识别，避免误报为"转发失败"
    if (streamCtx) streamCtx.aborted = true;
    if (log) log(`上游流式响应超时，已主动断开 (${streamActivityTimeout / 1000}秒无数据)`);
    gracefulSseClose(`上游流式响应超时 (${streamActivityTimeout / 1000}秒无数据)`);
    proxyRes.destroy();
    finish();
  }

  let activityTimer = setTimeout(onActivityTimeout, streamActivityTimeout);
  activityTimer.unref();

  function resetActivityTimer() {
    clearTimeout(activityTimer);
    activityTimer = setTimeout(onActivityTimeout, streamActivityTimeout);
    activityTimer.unref();
  }

  const sseHeaders = {
    'content-type':  'text/event-stream',
    'cache-control': 'no-cache',
    'connection':    'keep-alive',
  };
  if (clientFormat === 'anthropic') {
    sseHeaders['anthropic-version'] = '2023-06-01';
  }
  clientRes.writeHead(200, sseHeaders);

  // 相同格式：直接 pipe，保留完整 SSE 格式，打字机效果正常
  const isPassthrough =
    (providerFormat === 'anthropic' && clientFormat === 'anthropic') ||
    (providerFormat === 'openai'    && clientFormat === 'openai-chat');

  if (isPassthrough) {
    proxyRes.pipe(clientRes);

    proxyRes.on('data', resetActivityTimer);

    proxyRes.on('end', finish);

    // 上游流出错时也要释放队列 slot
    proxyRes.on('error', (err) => {
      if (!clientRes.writableEnded) clientRes.end();
      finish();
    });

    // 客户端断开时清理上游连接，释放队列 slot
    clientRes.on('close', () => {
      proxyRes.destroy();
      finish();
    });

    return;
  }

  // 不同格式：逐行解析并转换 SSE 事件
  const transform = createStreamTransformer(providerFormat, clientFormat);
  let buffer = '';

  proxyRes.on('data', (chunk) => {
    resetActivityTimer();
    buffer += chunk.toString('utf-8');
    const lines = buffer.split('\n');
    buffer = lines.pop();

    for (const line of lines) {
      const trimmed = line.trimEnd();
      if (!trimmed) { clientRes.write('\n'); continue; }
      for (const out of transform(trimmed)) {
        if (out) clientRes.write(out);
      }
    }
  });

  proxyRes.on('end', () => {
    if (buffer.trim()) {
      for (const out of transform(buffer.trim())) {
        if (out) clientRes.write(out);
      }
    }
    if (!clientRes.writableEnded) clientRes.end();
    finish();
  });

  proxyRes.on('error', (err) => {
    if (!clientRes.writableEnded) clientRes.end();
    finish();
  });

  // 客户端断开时清理上游连接，释放队列 slot
  clientRes.on('close', () => {
    proxyRes.destroy();
    finish();
  });
}

// ── 错误响应处理 ──────────────────────────────

function handleError(proxyRes, clientRes, statusCode, log) {
  let raw = '';
  proxyRes.on('data', c => raw += c);
  proxyRes.on('end', () => {
    log(`   错误响应: ${raw.slice(0, 300)}`);
    if (!clientRes.headersSent) {
      clientRes.writeHead(statusCode, { 'content-type': 'application/json' });
      clientRes.end(raw);
    }
  });
  // 上游错误响应流本身出错时，确保客户端能收到响应，及时回收 socket
  proxyRes.on('error', (err) => {
    log(`   错误响应流异常: ${err.message}`);
    if (!clientRes.headersSent) {
      clientRes.writeHead(502, { 'content-type': 'application/json' });
      clientRes.end(JSON.stringify({ error: { type: 'upstream_error', message: err.message } }));
    } else if (!clientRes.writableEnded) {
      clientRes.end();
    }
  });
}

File: go/internal/proxy/proxy.go
// Package proxy 对齐 Node 版 lib/proxy.js：转发请求到上游，处理流式/非流式响应。
//
// 相比 Node 版手搓 setTimeout + proxyRes.destroy + streamCtx 标志，Go 版用 context 统一管理
// 请求超时与活跃超时，流式转发用单独 goroutine，取消/超时通过 ctx 一处收口，不会有 slot 泄漏。
package proxy

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync/atomic"
	"time"

	"ai-gateway/internal/config"
	"ai-gateway/internal/converter"
	"ai-gateway/internal/httputil"
)

// LogFunc 日志函数，绑定了 reqId
type LogFunc func(format string, args ...any)

// Options 转发选项
type Options struct {
	ClientReq             *http.Request
	ClientRes             http.ResponseWriter
	UpstreamBody          []byte // 已转换好的上游请求体
	Provider              *config.Provider
	ClientFormat          string // anthropic | openai-chat | openai-responses
	OriginalModel         string
	IsStreaming           bool
	Log                   LogFunc
	StartTime             time.Time
	TimeoutMs             int
	HeaderTimeoutMs       int // 流式：等上游响应头的超时（毫秒）。为 0 时回退用 TimeoutMs，保证非直通调用方行为不变
	StreamActivityTimeout int
	HTTPClient            *http.Client
}

// Forward 执行转发。返回时表示响应已完成（或已失败），调用方可释放队列 slot。
func Forward(opts *Options) error {
	upstreamURL, upstreamPath := buildUpstreamURL(opts.Provider)

	// 请求超时用 context 控制；流式响应的活跃超时单独处理（见 handleStream）
	ctx, cancel := context.WithCancel(opts.ClientReq.Context())
	defer cancel()

	timeout := time.Duration(opts.TimeoutMs) * time.Millisecond

	if opts.IsStreaming {
		// 流式路径：对 Do 阶段加 header 响应超时，避免上游接受连接但不发字节时永久阻塞
		// 拿到响应头后切换到活跃超时控制（见 handleStream）
		// header 超时优先用 HeaderTimeoutMs（直通模式），为 0 时回退到整体 TimeoutMs
		headerTimeoutMs := opts.TimeoutMs
		if opts.HeaderTimeoutMs > 0 {
			headerTimeoutMs = opts.HeaderTimeoutMs
		}
		headerTimeout := time.Duration(headerTimeoutMs) * time.Millisecond
		doCtx, doCancel := context.WithTimeout(ctx, headerTimeout)
		req, err := http.NewRequestWithContext(doCtx, http.MethodPost, upstreamURL+upstreamPath, bytes.NewReader(opts.UpstreamBody))
		if err != nil {
			doCancel()
			writeJSONError(opts.ClientRes, http.StatusBadGateway, "proxy_error", err.Error())
			return err
		}
		setUpstreamHeaders(req, opts.Provider)

		resp, err := opts.HTTPClient.Do(req)
		doCancel() // 释放 header 超时 context，后续靠活跃超时
		if err != nil {
			opts.Log("转发失败: %s", err.Error())
			if errors.Is(err, context.DeadlineExceeded) {
				writeJSONError(opts.ClientRes, http.StatusGatewayTimeout, "timeout_error", fmt.Sprintf("上游响应头超时 (%d秒)", headerTimeoutMs/1000))
				return err
			}
			writeJSONError(opts.ClientRes, http.StatusBadGateway, "proxy_error", err.Error())
			return err
		}
		defer resp.Body.Close()

		elapsed := time.Since(opts.StartTime).Milliseconds()
		if resp.StatusCode >= 400 {
			opts.Log("← HTTP %d [%dms]", resp.StatusCode, elapsed)
			return handleError(resp, opts.ClientRes, opts.Log)
		}
		opts.Log("← %d [%dms]", resp.StatusCode, elapsed)
		return handleStream(ctx, cancel, resp, opts)
	}

	// 非流式：整体超时
	ctx, cancel2 := context.WithTimeout(ctx, timeout)
	defer cancel2()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, upstreamURL+upstreamPath, bytes.NewReader(opts.UpstreamBody))
	if err != nil {
		writeJSONError(opts.ClientRes, http.StatusBadGateway, "proxy_error", err.Error())
		return err
	}
	setUpstreamHeaders(req, opts.Provider)

	resp, err := opts.HTTPClient.Do(req)
	if err != nil {
		opts.Log("转发失败: %s", err.Error())
		if errors.Is(err, context.DeadlineExceeded) {
			writeJSONError(opts.ClientRes, http.StatusGatewayTimeout, "timeout_error", fmt.Sprintf("上游请求超时 (%d秒)", opts.TimeoutMs/1000))
			return err
		}
		writeJSONError(opts.ClientRes, http.StatusBadGateway, "proxy_error", err.Error())
		return err
	}
	defer resp.Body.Close()

	elapsed := time.Since(opts.StartTime).Milliseconds()
	if resp.StatusCode >= 400 {
		opts.Log("← HTTP %d [%dms]", resp.StatusCode, elapsed)
		return handleError(resp, opts.ClientRes, opts.Log)
	}
	opts.Log("← %d [%dms]", resp.StatusCode, elapsed)
	return handleResponse(resp, opts)
}

// handleResponse 非流式响应：读取整体，按格式转换后回写。
func handleResponse(resp *http.Response, opts *Options) error {
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		opts.Log("上游响应流错误: %s", err.Error())
		writeJSONError(opts.ClientRes, http.StatusBadGateway, "upstream_error", err.Error())
		return err
	}

	var data map[string]any
	if err := json.Unmarshal(raw, &data); err != nil {
		writeJSONError(opts.ClientRes, http.StatusBadGateway, "parse_error", "上游响应解析失败: "+err.Error())
		return err
	}

	var result map[string]any
	if opts.Provider.Format == "anthropic" {
		result = converter.ConvertAnthropicResponse(data, opts.ClientFormat, opts.OriginalModel)
	} else {
		result = converter.ConvertOpenAIChatResponse(data, opts.ClientFormat, opts.OriginalModel)
	}

	out, _ := json.Marshal(result)
	opts.ClientRes.Header().Set("content-type", "application/json")
	opts.ClientRes.WriteHeader(http.StatusOK)
	_, err = opts.ClientRes.Write(out)
	return err
}

// handleStream 流式响应：核心改善点。
// 用活跃超时计时器：每收到一块数据就重置；超时则 cancel ctx 断开上游，并向客户端补发合规 SSE 收尾事件。
func handleStream(ctx context.Context, cancel context.CancelFunc, resp *http.Response, opts *Options) error {
	flusher, ok := opts.ClientRes.(http.Flusher)
	if !ok {
		return fmt.Errorf("ResponseWriter 不支持 Flush，无法流式转发")
	}

	h := opts.ClientRes.Header()
	h.Set("content-type", "text/event-stream")
	h.Set("cache-control", "no-cache")
	h.Set("connection", "keep-alive")
	if opts.ClientFormat == "anthropic" {
		h.Set("anthropic-version", "2023-06-01")
	}
	opts.ClientRes.WriteHeader(http.StatusOK)
	flusher.Flush()

	activityTimeout := time.Duration(opts.StreamActivityTimeout) * time.Millisecond

	// 活跃超时计时器：到期 cancel ctx，从而中断下面的 Read
	timer := time.NewTimer(activityTimeout)
	defer timer.Stop()
	var timedOut atomic.Bool
	go func() {
		select {
		case <-timer.C:
			timedOut.Store(true)
			cancel() // 中断上游读取
		case <-ctx.Done():
		}
	}()

	// 相同格式直接透传字节流，保留完整 SSE 格式与打字机效果；不同格式逐行解析转换。
	isPassthrough := converter.IsPassthrough(opts.Provider.Format, opts.ClientFormat)

	if isPassthrough {
		buf := make([]byte, 16*1024)
		for {
			n, err := resp.Body.Read(buf)
			if n > 0 {
				timer.Reset(activityTimeout)
				if _, werr := opts.ClientRes.Write(buf[:n]); werr != nil {
					cancel()
					return werr
				}
				flusher.Flush()
			}
			if err != nil {
				return finishStream(err, timedOut.Load(), opts, flusher)
			}
		}
	}

	// 不同格式：逐行解析并经 transformer 转换（对齐 Node 的按 \n 切分逻辑）。
	// 手动按 \n 分行读取，无 bufio.Scanner 的 4MB 单行上限，避免大 SSE 事件被截断。
	transform := converter.NewStreamTransformer(opts.Provider.Format, opts.ClientFormat)
	readBuf := make([]byte, 64*1024)
	var lineBuf []byte
	for {
		n, readErr := resp.Body.Read(readBuf)
		if n > 0 {
			timer.Reset(activityTimeout)
			chunk := readBuf[:n]
			for {
				idx := bytes.IndexByte(chunk, '\n')
				if idx < 0 {
					// 无完整行，剩余数据暂存到 lineBuf
					lineBuf = append(lineBuf, chunk...)
					break
				}
				line := string(append(lineBuf, chunk[:idx]...))
				lineBuf = lineBuf[:0]
				chunk = chunk[idx+1:]
				line = strings.TrimRight(line, "\r")
				if line == "" {
					if _, werr := opts.ClientRes.Write([]byte("\n")); werr != nil {
						cancel()
						return werr
					}
					flusher.Flush()
					continue
				}
				for _, out := range transform.Transform(line) {
					if out == "" {
						continue
					}
					if _, werr := opts.ClientRes.Write([]byte(out)); werr != nil {
						cancel()
						return werr
					}
				}
				flusher.Flush()
			}
		}
		if readErr != nil {
			// 流结束：lineBuf 中可能有不以 \n 结尾的最后一行
			if len(lineBuf) > 0 {
				line := strings.TrimRight(string(lineBuf), "\r")
				if line != "" {
					for _, out := range transform.Transform(line) {
						if out != "" {
							opts.ClientRes.Write([]byte(out))
						}
					}
					flusher.Flush()
				}
			}
			return finishStream(readErr, timedOut.Load(), opts, flusher)
		}
	}
}

// finishStream 统一处理流结束：正常 EOF / 活跃超时 / 客户端断开。
func finishStream(err error, timedOut bool, opts *Options, flusher http.Flusher) error {
	if err == nil || err == io.EOF {
		return nil // 正常结束
	}
	if timedOut {
		opts.Log("上游流式响应超时，已主动断开 (%d秒无数据)", opts.StreamActivityTimeout/1000)
		gracefulSSEClose(opts.ClientRes, flusher, opts.ClientFormat, opts.StreamActivityTimeout)
		return nil
	}
	// context.Canceled：客户端主动断开连接，属正常结束而非异常
	if errors.Is(err, context.Canceled) {
		opts.Log("客户端断开连接")
		return nil
	}
	return err // 其它读取错误
}

// gracefulSSEClose 向客户端补发合规 SSE 收尾事件，避免截断流（对齐 Node 版改动）。
func gracefulSSEClose(w http.ResponseWriter, f http.Flusher, clientFormat string, timeoutMs int) {
	if clientFormat == "anthropic" {
		msg := fmt.Sprintf("上游流式响应超时 (%d秒无数据)", timeoutMs/1000)
		payload, _ := json.Marshal(map[string]any{
			"type":  "error",
			"error": map[string]any{"type": "timeout_error", "message": msg},
		})
		fmt.Fprintf(w, "event: error\ndata: %s\n\n", payload)
	} else {
		fmt.Fprint(w, "data: [DONE]\n\n")
	}
	f.Flush()
}

// handleError 上游错误响应：原样回传状态码与响应体。
func handleError(resp *http.Response, w http.ResponseWriter, log LogFunc) error {
	raw, _ := io.ReadAll(resp.Body)
	if len(raw) > 300 {
		log("   错误响应: %s", raw[:300])
	} else {
		log("   错误响应: %s", raw)
	}
	w.Header().Set("content-type", "application/json")
	w.WriteHeader(resp.StatusCode)
	_, err := w.Write(raw)
	return err
}

// writeJSONError 代理内部使用 httputil 共享实现。
func writeJSONError(w http.ResponseWriter, status int, errType, msg string) {
	httputil.WriteJSONError(w, status, errType, msg)
}
